src/debugger/debugging.ts

```diff
@@ -0,0 +1,88 @@
+
+export class TLADebugAdapterServerDescriptorFactory implements vscode.DebugAdapterDescriptorFactory {
+
+    createDebugAdapterDescriptor(session: vscode.DebugSession, executable: vscode.DebugAdapterExecutable | undefined):
+        vscode.ProviderResult<vscode.DebugAdapterDescriptor> {
+        return new vscode.DebugAdapterServer(session.configuration['port']);
+    }
+}
+
+/**
+ * Attaches the DAP front-end to an already running TLC debugger.
+ */
+export async function debugSpec(
+    resource: vscode.Uri | undefined,
+    diagnostic: vscode.DiagnosticCollection,
+    context: vscode.ExtensionContext
+): Promise<void> {
+    let targetResource = resource;
+    if (!targetResource && vscode.window.activeTextEditor) {
+        // Since this command is registered as a button on the editor menu, I don't
+        // think this branch is ever taken.  It's here because the DAP example has it.
+        targetResource = vscode.window.activeTextEditor.document.uri;
+    }
+    if (targetResource) {
+        // Attaching to a separately launched TLC leaves the result view (webview) empty.
+        // However, TLC sends its output via the DAP Output event
+        // (https://microsoft.github.io/debug-adapter-protocol/specification#Events_Output)
+        // to VSCode.  Somebody has to figure out how to wire this up. In the meantime,
+        // users have to manually "parse" the output on VSCode's DebugConsole. Unfortunately,
+        // it is TLC's '-tool' output containing "@!@!@" markers around each message.
+        vscode.debug.startDebugging(undefined, {
+            type: 'tlaplus',
+            name: 'Debug Spec',
+            request: 'launch',
+            program: targetResource.fsPath,
+            port: '4712'
+        });
+    }
+}
+
+/**
+ * Runs TLC in debugger mode and attaches the DAP front-end.
+ */
+export async function checkAndDebugSpec(
+    resource: vscode.Uri | undefined,
+    diagnostic: vscode.DiagnosticCollection,
+    context: vscode.ExtensionContext
+): Promise<void> {
+    let targetResource = resource;
+    if (!targetResource && vscode.window.activeTextEditor) {
+        // Since this command is registered as a button on the editor menu, I don't
+        // think this branch is ever taken.  It's here because the DAP example has it.
+        targetResource = vscode.window.activeTextEditor.document.uri;
+    }
+    if (targetResource) {
+        const specFiles = await getSpecFiles(targetResource);
+        if (!specFiles) {
+            return;
+        }
+        // Randomly select a port on which we request the debugger to listen
+        const port = Math.floor(Math.random() * (64510 - 1025 + 1)) + 1025;
+        // false => Don't open the result view, it's empty anyway (see above).
+        // Don't await doCheckModel because it only returns after TLC terminates.
+        doCheckModel(specFiles, false, context, diagnostic, false, ['-debugger', `port=${port}`]);
```

Comment:
I don't quite understand why "it's empty anyway". Web view should work fine here. Could you, please, amplify on why you pass `false`?

Reply:
I don't remember exactly, but that commented might be outdated (I can't test it right now).  Anyway, when a user launches the TLA+ debugger, she doesn't care about the TLC statistics, coverage, ..., the web view in general.  She wants to see the spec editor.  If she occasionally indeed wants to see stats/coverage, the display model panel command is a click away.

Reply:
I agree that it's not necessary to open the web view on debugger launch. Though, it may come handy if checking ends with an error, and it won't be a click away by default, it'll require running the checker again.

But I'm just a bit confused by the first comment, I think it's incorrect.